storm-core/src/jvm/org/apache/storm/scheduler/IsolationScheduler.java

```diff
@@ -0,0 +1,417 @@
+                distribution.put(value, newValue);
+            }
+        }
+    }
+
+    private Map<String, Set<WorkerSlot>> hostUsedSlots(Cluster cluster) {
+        Collection<WorkerSlot> usedSlots = cluster.getUsedSlots();
+        Map<String, Set<WorkerSlot>> hostUsedSlots = new HashMap<String, Set<WorkerSlot>>();
+        for (WorkerSlot slot : usedSlots) {
+            String host = cluster.getHost(slot.getNodeId());
+            Set<WorkerSlot> slots = hostUsedSlots.get(host);
+            if (slots == null) {
+                slots = new HashSet<WorkerSlot>();
+                hostUsedSlots.put(host, slots);
+            }
+            slots.add(slot);
+        }
+        return hostUsedSlots;
+    }
+
+    // returns list of list of slots, reverse sorted by number of slots
+    private LinkedList<HostAssignableSlots> hostAssignableSlots(Cluster cluster) {
+        List<WorkerSlot> assignableSlots = cluster.getAssignableSlots();
+        Map<String, List<WorkerSlot>> hostAssignableSlots = new HashMap<String, List<WorkerSlot>>();
+        for (WorkerSlot slot : assignableSlots) {
+            String host = cluster.getHost(slot.getNodeId());
+            List<WorkerSlot> slots = hostAssignableSlots.get(host);
+            if (slots == null) {
+                slots = new ArrayList<WorkerSlot>();
+                hostAssignableSlots.put(host, slots);
+            }
+            slots.add(slot);
+        }
+        List<HostAssignableSlots> sortHostAssignSlots = new ArrayList<HostAssignableSlots>();
+        for (Map.Entry<String, List<WorkerSlot>> entry : hostAssignableSlots.entrySet()) {
+            sortHostAssignSlots.add(new HostAssignableSlots(entry.getKey(), entry.getValue()));
+        }
+        Collections.sort(sortHostAssignSlots, new Comparator<HostAssignableSlots>() {
+                    @Override
+                    public int compare(HostAssignableSlots o1, HostAssignableSlots o2) {
+                        return o2.getWorkerSlots().size() - o1.getWorkerSlots().size();
+                    }
+                });
+        Collections.shuffle(sortHostAssignSlots);
+
+        return new LinkedList<HostAssignableSlots>(sortHostAssignSlots);
+    }
+
+    private List<Integer> distributionSortedAmts(Map<Integer, Integer> distributions) {
+        List<Integer> sorts = new ArrayList<Integer>();
+        for (Map.Entry<Integer, Integer> entry : distributions.entrySet()) {
+            int workers = entry.getKey();
+            int machines = entry.getValue();
+            for (int i = 0; i < machines; i++) {
+                sorts.add(workers);
+            }
+        }
+        Collections.sort(sorts, new Comparator<Integer>() {
+            @Override
+            public int compare(Integer o1, Integer o2) {
+                return o1.intValue() - o2.intValue();
+            }
+        });
+
+        return sorts;
+    }
+
+    private Set<String> allocatedTopologies(Map<String, Set<Set<ExecutorDetails>>> workerSpecs) {
```

Comment:
minor: topologyToWorkerSpecs might be easier to understand since workerSpecs is the value part of map.